File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);

        vector<vector<string>> answer;
        int strs_len = strs.size();
        unordered_map<string, vector<string>> map;

        for (const string& str : strs) {
            string sorted_str = str;
            sort(sorted_str.begin(), sorted_str.end());
            map[sorted_str].push_back(str);
        }

        for (const auto& ans : map) {
            answer.push_back(ans.second);
        }
        
        return answer;
    }
};

File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, List<String>> map = new HashMap<>();

        for (String str : strs) {
            char[] charArr = str.toCharArray();
            Arrays.sort(charArr);

            String key = new String(charArr);

            if (map.containsKey(key)) {
                map.get(key).add(str);
            }
            else {
                // map.computeIfAbsent(key, k -> new ArrayList<>()).add(str);
                List<String> list = new ArrayList<>();
                list.add(str);
                map.put(key, list);
            }
        }

        return new ArrayList<>(map.values());
    }
}